src/dali/plugin/input/rest/kucoin_futures.py

```diff
@@ -0,0 +1,288 @@
+from dali.intra_transaction import IntraTransaction
+from dali.out_transaction import OutTransaction
+
+_AMOUNT: str = "amount"
+_CURRENCY: str = "currency"
+_DATA: str = "data"
+_DATA_LIST: str = "dataList"
+_DEPOSIT: str = "Deposit"
+_END_AT: str = "endAt"
+_HAS_MORE: str = "hasMore"
+_OFFSET: str = "offset"
+_REALISED_PNL: str = "RealisedPNL"
+_START_AT: str = "startAt"
+_TIME: str = "time"
+_TRANSFER_IN: str = "TransferIn"
+_TRANSFER_OUT: str = "TransferOut"
+_TS_INCREMENT: int = 86400000
+_TYPE: str = "type"
+_WITHDRAWAL: str = "Withdrawal"
+
+class InputPlugin(AbstractCcxtInputPlugin):
+
+    __EXCHANGE_NAME: str = "kucoin_futures"
+    __PLUGIN_NAME: str = "kucoin_futures"
+    __DEFAULT_THREAD_COUNT: int = 1
+
+    def __init__(
+        self, account_holder: str, api_key: str, api_secret: str, api_passphrase: str, native_fiat: Optional[str] = None, thread_count: Optional[int] = None
+    ) -> None:
+        self.__api_key = api_key
+        self.__api_secret = api_secret
+        self.__api_passphrase = api_passphrase
+
+        super().__init__(account_holder, datetime(2019, 2, 17, 0, 0, 0, 0), native_fiat, thread_count)
+
+    def _initialize_client(self) -> kucoinfutures:
+        return kucoinfutures(
+            {
+                "apiKey": self.__api_key,
+                "secret": self.__api_secret,
+                "password": self.__api_passphrase,
+                "enableRateLimit": True,
+            }
+        )
+
+    def exchange_name(self) -> str:
+        return self.__EXCHANGE_NAME
+
+    def plugin_name(self) -> str:
+        return self.__PLUGIN_NAME
+
+    @property
+    def _client(self) -> kucoinfutures:
+        super_client: Exchange = super()._client
+        if not isinstance(super_client, kucoinfutures):
+            raise TypeError("super_client is not of type kucoinfutures")
+
+        return super_client
+
+    # using kucoin transaction history api instead. (implicit api)
+    def _get_process_deposits_pagination_detail_set(self) -> Optional[AbstractPaginationDetailSet]:
+        return None
+
+    def _get_process_withdrawals_pagination_detail_set(self) -> Optional[AbstractPaginationDetailSet]:
+        return None
+
+    def _get_process_trades_pagination_detail_set(self) -> Optional[AbstractPaginationDetailSet]:
+        return None
+
+    def _process_gains(
+        self,
+        in_transactions: List[InTransaction],
+        out_transactions: List[OutTransaction],
+    ) -> None:
+        pass
+
+    def _fetch_data(
+        self,
+        start_at: int,
+        end_at: int,
+    ) -> List[Dict[str, Any]]:
+
+        offset: int = 1
+        retries: int = 0
+        has_more: bool = False
+
+        ledger_transactions: Dict[str, Any] = {}
+        items: List[Dict[str, Any]] = []
+
+        while (start_at <= end_at) and retries < 4:
+            try:
+                ledger_transactions = self._client.futuresPrivateGetTransactionHistory({_START_AT: start_at, _END_AT: start_at + _TS_INCREMENT})
```

Comment:
Is there a reason why you are using the implicit api? There is a unified function for this `fetch_my_trades()` that the abstract plugin implements. You simply need to declare the `pagination_detail_set` for it.

All of the unified functions are wrapped to catch exceptions and are much more robust. Were you having issues with them? I'm working on some fixes now.

If you want to process the transactions differently, you can overload `_process_buy_and_sell()`

Reply:
I initially also tried to use fetch_my_trades. Except it only contains transactions fees and no funding fees. So If I were to use fetch_my_trade, I also need to call https://docs.kucoin.com/futures/#get-funding-history. Kucoin is rate limited by ip/account, not by endpoints. Funding history requires each future type as a parameter and there are I believe at least 30 different types. each call takes around 15 min to finish and fetch full transactions. So just to get complete trades, it would take 15 * 30 min at the minimum.

There could be optimizations here and there and try to reduce that time, but Kucoin futures also has transaction history endpoint which basically returns each trade's computed pnl including transaction fees & funding fees and I thought it would be simpler to just call transaction history endpoint.